Low level API support for Objective-C interop.

AaronRobinsonMSFT opened the pull request:
Implements approved API in https://github.com/dotnet/runtime/issues/44659#issuecomment-829495419.
Add support for `HNDTYPE_REFCOUNTED` GC handle type under specific `FEATURE_` defines.

Fixes https://github.com/dotnet/runtime/issues/44659

/cc @VSadov @Maoni0 @jkotas @jkoritzinsky @elinor-fung @rolfbjarne @lambdageek @davidwrighton @janvorli

jkoritzinsky committed:
Add initial infra for Objective-C interop support.

jkoritzinsky committed:
Stub out API and base directory for ObjectiveC interop tests (#47438)

elinor-fung committed:
Fix for building without FeatureObjcWrappers (#47445)

AaronRobinsonMSFT committed:
Connect VM to InteropLib (#47504)

* Add ComWrappers CMake feature flag
Split interoplibinterface for comwrappers and objc
Create pattern for QCalls

* Feedback and SAL.

AaronRobinsonMSFT committed:
Merge pull request #47507 from AaronRobinsonMSFT/objc_merge

Merge in latest .NET 6 development branch

elinor-fung committed:
Include handling of HNDTYPE_REFCOUNTED when FEATURE_OBJCWRAPPERS is defined (#47534)

elinor-fung committed:
Allow overriding of Objective-C message send P/Invokes (#47721)

elinor-fung committed:
Rename FEATURE_OBJCWRAPPERS to FEATURE_OBJCBRIDGE

Remove System.Runtime.InteropServices.ObjectiveC.Wrappers and associated classes

elinor-fung committed:
Objc -> ObjC

AaronRobinsonMSFT committed:
Add managed side of reference tracker API.

elinor-fung committed:
Rename FEATURE_OBJCWRAPPERS to FEATURE_OBJCBRIDGE (#47782)

Remove System.Runtime.InteropServices.ObjectiveC.Wrappers and associated classes

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/feature/objc-interop' into reftracker

AaronRobinsonMSFT committed:
Add new flag to MethodTable for types with TrackedNativeReferenceAttribute.
Discover attribute during type load and set bit.
Update EagerFinalizer callout to query bit.

AaronRobinsonMSFT committed:
First pass at API complete.

AaronRobinsonMSFT committed:
Error on incorrect API usage.

AaronRobinsonMSFT committed:
Add testing for API

AaronRobinsonMSFT committed:
Add test for creating new handles for the same object.

AaronRobinsonMSFT committed:
Feedback

AaronRobinsonMSFT committed:
Update src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/ObjectiveC/Bridge.cs

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

AaronRobinsonMSFT committed:
Remove unnecessary static lib

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/master' into reftracker

AaronRobinsonMSFT committed:
Merge pull request #47832 from AaronRobinsonMSFT/reftracker

Expose the refcounted GCHandle from the Objective-C interop layer

AaronRobinsonMSFT committed:
Enable a "pending exception" mechanism for msgSend functions.

AaronRobinsonMSFT committed:
Fix issues build/run on macOS

AaronRobinsonMSFT committed:
Testing for pending exceptions API

AaronRobinsonMSFT committed:
Feedback

AaronRobinsonMSFT committed:
Missed comment

AaronRobinsonMSFT committed:
Bridge API for propagating managed exception to native.
- Only for OSX_64

AaronRobinsonMSFT committed:
Remove assembly stub.

AaronRobinsonMSFT committed:
Pass a RuntimeMethodHandle to the event handler for the last method on
the stack prior to control leaving the runtime.

AaronRobinsonMSFT committed:
Fail fast if a handler throws an exception.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into managed_native_exception

AaronRobinsonMSFT committed:
Temporary ARM64 fix for exception propagation.

AaronRobinsonMSFT committed:
Update for ARM64

AaronRobinsonMSFT committed:
Do the right thing for both x64 and ARM64 for exception prop.
Collapse some duplicate code paths.

AaronRobinsonMSFT committed:
Add support for ARM.
Change compile time error to failfast for non-implemented platforms.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into managed_native_exception

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into managed_native_exception

AaronRobinsonMSFT committed:
Add new VM/GC callbacks
  - RefCountedHandleCallbacksBefore
  - RefCountedHandleCallbacksAfter

AaronRobinsonMSFT committed:
Update GC callout for RefCounted handled begin/end pair.
Update tests to handle new API.
  - no longer pass int to callback for begin/end

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into managed_native_exception

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into feature/objc-interop-prototype

AaronRobinsonMSFT committed:
Merge in main and fix up API usage.

AaronRobinsonMSFT committed:
[BUILD-ONLY] Update API surface area.

AaronRobinsonMSFT committed:
Missed building ref assemblies.

AaronRobinsonMSFT committed:
tests and libraries tests updates.

AaronRobinsonMSFT committed:
Product and tests update to new names.

AaronRobinsonMSFT committed:
Fix comments on class.

AaronRobinsonMSFT committed:
Fix sign issue.
Always include the interoplibinterface shared component.

AaronRobinsonMSFT committed:
Fix bad merge.

VSadov commented on src/coreclr/gc/objecthandle.cpp:
```diff
@@ -1065,7 +1065,9 @@ void Ref_TraceNormalRoots(uint32_t condemned, uint32_t maxgen, ScanContext* sc,
     TraceVariableHandles(PromoteObject, uintptr_t(sc), uintptr_t(fn), VHT_STRONG, condemned, maxgen, flags);
 
-#if defined(FEATURE_COMINTEROP) || defined(FEATURE_REDHAWK)
+    GCToEEInterface::RefCountedHandleCallbacksBefore();
```
This does not need to be called when `sc->concurrent == true`. Same applies to the "After" part.
When handles are scanned in the background (concurrently with managed threads), ref-counted handles are not scanned, so callbacks will be redundant.
See the condition at the line 1072.

VSadov replied:
What kind of code is runs in this Begin/After callback? Do we have an example?

AaronRobinsonMSFT replied:
@VSadov Thanks for the clarity. I may have misunderstood @Maoni0 when this location was suggested, will let her clarify but I applied your suggestion and updated the naming on the interop side.

>What kind of code runs in this Begin/After callback? Do we have an example?

Sure. An example is in the test which isn't how the product will do this but the idea is to let the other system know when those handles are being examined and tracking that state.

```cpp
    std::atomic_bool state = { false };

     void STDMETHODCALLTYPE BeginEndCb()
    {
#ifdef REF_TRACKER_DEBUG_CALLBACKS
        ::printf("BeginEndCb: %s -> %s\n",
            state ? "true" : "false",
            !state ? "true" : "false");
#endif // REF_TRACKER_DEBUG_CALLBACKS

        state = !state;
    }
```

VSadov replied:
My concern is that on server GC this will end up being called N times in parallel where N is the number of heaps, which is typically #proc. So the `BeginEnd` callback will see N begin calls and N end calls in some random order. It looks like your callback is not expecting such scenario.

It could get worse if the real callback takes locks as that would have serializing effect on GC - it will force the handle marking to happen serially, one heap after another.

I think these Begin/End calls should happen only once per GC (background or regular) -
-  one call before marking handles and one after in a regular (nonconcurrent) mark, and 
-  one call before marking handles and one after in a nonconcurrent mark phase of a background GC. 

In the second case the callbacks should happen when worker threads are joined, so that only one worker thread makes calls, not all of them. 
I think we have convenient locations in GC to put these callbacks, but you should check with @Maoni0 if that makes sense and where exactly the callbacks should be placed.

AaronRobinsonMSFT committed:
Fix library tests.

AaronRobinsonMSFT committed:
Review feedback and name updates on the interop side.

AaronRobinsonMSFT committed:
Fix linker issue on Windows.

AaronRobinsonMSFT committed:
Disable the new ObjectiveCMarshal API

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into feature/objc-interop-prototype

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into feature/objc-interop-prototype

AaronRobinsonMSFT committed:
Fix merge of upstream/main.

AaronRobinsonMSFT committed:
Update GCEE interface to support GCServer scenario for Objective-C.
The new callbacks now fire when before the GC scans handles and is
non-concurrent.

AaronRobinsonMSFT commented:
/azp run runtime

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23683,6 +23683,7 @@ void gc_heap::mark_phase (int condemned_gen_number, BOOL mark_only_p)
         grow_mark_list_piece();
 #endif //USE_REGIONS
 
+        GCToEEInterface::BeforeGcScanRoots(max_generation, /* is_bgc */ false, /* is_concurrent */ false);
```
this should be `condemned_gen_number` instead of max_generation as we are not necessarily collecting max_generation

AaronRobinsonMSFT replied:
Boo. Thanks. I completely missed this.

Maoni0 commented on src/coreclr/vm/gcenv.ee.cpp:
```diff
@@ -49,6 +49,24 @@ VOID GCToEEInterface::SyncBlockCacheWeakPtrScan(HANDLESCANPROC scanProc, uintptr
+
+#ifdef VERIFY_HEAP
+    // Validate byrefs pinned by IL stubs since the last GC.
+    StubHelpers::ProcessByrefValidationList();
```
do you want to only call this when is_bgc is true? this is already called in GcStartWork which means for blocking GCs it would have just been called at the beginning of this GC.

AaronRobinsonMSFT replied:
Doh. Good call. I will update this to be what it was.

Maoni0 reviewed (approved):
the rest LGTM!

AaronRobinsonMSFT committed:
Feedback on GC callbacks

AaronRobinsonMSFT committed:
Clarify documentation.

janvorli reviewed (commented):
The exception handling stuff looks good modulo the comment.

VSadov reviewed (approved):
PInvoke overriders and GC callbacks LGTM.

AaronRobinsonMSFT commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -4438,13 +4488,25 @@ VOID UnwindManagedExceptionPass2(PAL_SEHException& ex, CONTEXT* unwindStartConte
                 // obsolete after the unwind of the native frames completes.
                 ExceptionTracker* pTracker = GetThread()->GetExceptionState()->GetCurrentExceptionTracker();
-                pTracker->CleanupBeforeNativeFramesUnwind();
+                if (pTracker != NULL)
```
Still investigating this. So far, I've been able to confirm the `ExceptionTracker` instance is being NULL'd out during the last call to `ProcessCLRException` in `UnwindManagedExceptionPass2`.

AaronRobinsonMSFT replied:
@janvorli I think this is actually okay. The `ExceptionTracker` instance is being cleaned up in `ProcessCLRException` because of the "invalid Reverse P/Invoke" logic at 

https://github.com/dotnet/runtime/blob/82c917aa6119c553c4fa093265c52f34476e7e13/src/coreclr/vm/exceptionhandling.cpp#L1229-L1235

I believe this is the cleanup we want. Should I assert something in this area?

AaronRobinsonMSFT replied:
Added comment as to why this is okay.

AaronRobinsonMSFT committed:
Comment added for ExceptionTracker non-NULL condition.

filipnavara commented:
Thanks! Happy to see this done! 🎉
